In [1017]:
import pandas as pd

# from utlis import rm_units_and_conv_int
import utlis

import importlib
importlib.reload(utlis)

<module 'utlis' from 'd:\\1. Projects\\DNA\\Portfolio\\car_market_analysis\\utlis.py'>

### **Merging all datasets into one**

Make sure to uncomment cells in this section in order to create a dataset that contains all data from data scraping.

In [1018]:
# from os import listdir

# PATH = 'data_scraping/data/datasets/'

# # Concatenating all datasets
# for i, ds in enumerate(listdir(PATH)):
#     if i == 0:
#         data = pd.read_csv(f'{PATH}{ds}')
#     else:
#         data = pd.concat([data, pd.read_csv(f'{PATH}{ds}')])

# print(data.shape)
# data.head()

In [1019]:
# # Translating column names from polish to english
# colnames = {
#     'Marka pojazdu': 'brand',
#     'Model pojazdu': 'model',
#     'Rok produkcji': 'year',
#     'Przebieg': 'mileage',
#     'Pojemność skokowa': 'engine_capacity',
#     'Rodzaj paliwa': 'fuel',
#     'Moc': 'power',
#     'Skrzynia biegów': 'gearbox',
#     'Napęd': 'drive',
#     'Spalanie W Mieście': 'city_fuel_consumption',
#     'Typ nadwozia': 'body',
#     'Liczba drzwi': 'doors',
#     'Liczba miejsc': 'seats',
#     'Kolor': 'color',
#     'Kraj pochodzenia': 'origin_country',
#     'Bezwypadkowy': 'accident_free',
#     'Serwisowany w ASO': 'serviced_at_ASO',
#     'Stan': 'condition',
#     'Cena': 'price'
# }
# data.rename(columns=colnames, inplace=True)

# data.head()

In [1020]:
# # Save all data in one file
# data.to_csv('data_scraping/data/all_data.csv', index=False)

# **Data cleaning**

In [1021]:
data = pd.read_csv('data_scraping/data/all_data.csv')
data.head()

brand model    year     mileage engine_capacity     fuel   power  \
0  Audi   SQ7  2017.0  142 800 km       3 956 cm3   Diesel  435 KM   
1  Audi    A3  2010.0  236 000 km       1 395 cm3  Benzyna  125 KM   
2  Audi    A4  2018.0   84 500 km       1 968 cm3   Diesel  190 KM   
3  Audi    Q5  2018.0   38 000 km       1 984 cm3  Benzyna  252 KM   
4  Audi    A7  2016.0   74 885 km       2 995 cm3  Benzyna  333 KM   

        gearbox             drive city_fuel_consumption     body  doors  \
0  Automatyczna       4x4 (stały)             8 l/100km      SUV    5.0   
1      Manualna  Na przednie koła                   NaN  Kompakt    5.0   
2  Automatyczna       4x4 (stały)             5 l/100km    Kombi    5.0   
3  Automatyczna       4x4 (stały)             8 l/100km      SUV    5.0   
4  Automatyczna       4x4 (stały)            10 l/100km  Kompakt    5.0   

   seats   color origin_country accident_free serviced_at_ASO condition  \
0    7.0   Szary         Polska           Tak             Tak   Używane   
1    5.0  Czarny            NaN           Tak             Tak   Używane   
2    5.0   Szary            NaN           NaN             NaN   Używane   
3    5.0   Szary            NaN           NaN             NaN   Używane   
4    5.0   Biały            NaN           NaN             NaN   Używane   

                price  
0  269 000        PLN  
1   29 900        PLN  
2   60 000        PLN  
3   75 000        PLN  
4   75 000        PLN

In [1022]:
data.shape

(134057, 19)

This dataset comprises a vast collection of 134,057 observations specifically centered around cars. With 18 informative columns, it offers a diverse range of details pertaining to each vehicle. These encompass essential information such as the brand, model, production year, mileage, and fuel type. 

## **Missing values**


In [1023]:
# How many missing values are there in each column
na_count = data.isna().sum()
na_precentage = (na_count / data.shape[0] * 100).round(2)

na_df = pd.DataFrame({'na_count': na_count, 'na_precentage': na_precentage})
na_df.sort_values(by='na_precentage', ascending=False)

na_count  na_precentage
serviced_at_ASO           73520          54.84
accident_free             58608          43.72
city_fuel_consumption     54864          40.93
origin_country            45703          34.09
drive                     17584          13.12
seats                      8409           6.27
engine_capacity            2441           1.82
mileage                    1093           0.82
doors                       735           0.55
power                        83           0.06
gearbox                      62           0.05
color                         1           0.00
condition                     0           0.00
brand                         1           0.00
body                          1           0.00
model                         1           0.00
fuel                          1           0.00
year                          1           0.00
price                         0           0.00

In this dataset, a notable observation is that the majority of columns contain missing values. However, an encouraging aspect is that the columns with missing values appear to be of lesser significance (such as whether the car was serviced at ASO, its accident-free status, and city fuel consumption). On the other hand, the essential data elements, including price, model, brand, and production year, exhibit minimal or no missing values at all. This characteristic suggests that the crucial information crucial for analysis is largely intact, allowing for comprehensive examination and insights into the dataset.

In [1024]:
# There is one row with majority of values missing, so I filtered it out. It may be caused by some error during scraping
data[data['brand'].isna()]

brand model  year mileage engine_capacity fuel power gearbox drive  \
66878   NaN   NaN   NaN     NaN             NaN  NaN   NaN     NaN   NaN   

      city_fuel_consumption body  doors  seats color origin_country  \
66878                   NaN  NaN    NaN    NaN   NaN            NaN   

      accident_free serviced_at_ASO condition              price  
66878           NaN             NaN   Używane  55 900        PLN

In [1025]:
data = data[data['brand'].notna()]

## **Data types**

In [1026]:
data.head()

brand model    year     mileage engine_capacity     fuel   power  \
0  Audi   SQ7  2017.0  142 800 km       3 956 cm3   Diesel  435 KM   
1  Audi    A3  2010.0  236 000 km       1 395 cm3  Benzyna  125 KM   
2  Audi    A4  2018.0   84 500 km       1 968 cm3   Diesel  190 KM   
3  Audi    Q5  2018.0   38 000 km       1 984 cm3  Benzyna  252 KM   
4  Audi    A7  2016.0   74 885 km       2 995 cm3  Benzyna  333 KM   

        gearbox             drive city_fuel_consumption     body  doors  \
0  Automatyczna       4x4 (stały)             8 l/100km      SUV    5.0   
1      Manualna  Na przednie koła                   NaN  Kompakt    5.0   
2  Automatyczna       4x4 (stały)             5 l/100km    Kombi    5.0   
3  Automatyczna       4x4 (stały)             8 l/100km      SUV    5.0   
4  Automatyczna       4x4 (stały)            10 l/100km  Kompakt    5.0   

   seats   color origin_country accident_free serviced_at_ASO condition  \
0    7.0   Szary         Polska           Tak             Tak   Używane   
1    5.0  Czarny            NaN           Tak             Tak   Używane   
2    5.0   Szary            NaN           NaN             NaN   Używane   
3    5.0   Szary            NaN           NaN             NaN   Używane   
4    5.0   Biały            NaN           NaN             NaN   Używane   

                price  
0  269 000        PLN  
1   29 900        PLN  
2   60 000        PLN  
3   75 000        PLN  
4   75 000        PLN

In [1027]:
data.dtypes

brand                     object
model                     object
year                     float64
mileage                   object
engine_capacity           object
fuel                      object
power                     object
gearbox                   object
drive                     object
city_fuel_consumption     object
body                      object
doors                    float64
seats                    float64
color                     object
origin_country            object
accident_free             object
serviced_at_ASO           object
condition                 object
price                     object
dtype: object

The dataset contains a few data type errors that need to be addressed.  
Because there are missing values in the dataset I convert all numeric columns to float.
* **mileage** should be `float`
* **engine_capacity** should be `float`
* **power** should be `float`
* **city_fuel_consumption** should be `float`
* **price** should be `float`

### **mileage** 

In [1028]:
data['mileage'].head()

0    142 800 km
1    236 000 km
2     84 500 km
3     38 000 km
4     74 885 km
Name: mileage, dtype: object

In [1029]:
data['mileage'] = utlis.rm_units_and_conv_dtype(data['mileage'], 'km', 'float')

data['mileage'].head()

0    142800.0
1    236000.0
2     84500.0
3     38000.0
4     74885.0
Name: mileage, dtype: float64

### **engine_capacity** 

In [1030]:
data['engine_capacity'].head()

0    3 956 cm3
1    1 395 cm3
2    1 968 cm3
3    1 984 cm3
4    2 995 cm3
Name: engine_capacity, dtype: object

In [1031]:
data['engine_capacity'] = utlis.rm_units_and_conv_dtype(data['engine_capacity'], 'cm3', 'float')

data['engine_capacity'].head()

0    3956.0
1    1395.0
2    1968.0
3    1984.0
4    2995.0
Name: engine_capacity, dtype: float64

### **power** 

In [1032]:
data['power'].head()

0    435 KM
1    125 KM
2    190 KM
3    252 KM
4    333 KM
Name: power, dtype: object

In [1033]:
data['power'] = utlis.rm_units_and_conv_dtype(data['power'], 'KM', 'float')

data['power'].head()

0    435.0
1    125.0
2    190.0
3    252.0
4    333.0
Name: power, dtype: float64

### **city_fuel_consumption** 

In [1034]:
data['city_fuel_consumption'].head()

0     8 l/100km
1           NaN
2     5 l/100km
3     8 l/100km
4    10 l/100km
Name: city_fuel_consumption, dtype: object

In [1035]:
data['city_fuel_consumption'] = utlis.rm_units_and_conv_dtype(data['city_fuel_consumption'], 'l/100km', 'float')

data['city_fuel_consumption'].head()

0     8.0
1     NaN
2     5.0
3     8.0
4    10.0
Name: city_fuel_consumption, dtype: float64

### **price** 

In [1036]:
data['price'].head()

0    269 000        PLN
1     29 900        PLN
2     60 000        PLN
3     75 000        PLN
4     75 000        PLN
Name: price, dtype: object

In [1037]:
# There are some prices with "PLN" and some with "EUR", so for this column different approach is needed

# Extracting the value and currency from the price column
data['value'] = data['price'].map(lambda x: float(''.join(x.split()[:-1]).replace(',', '.')))
data['currency'] = data['price'].map(lambda x: x.split()[-1])

data.head()

brand model    year   mileage  engine_capacity     fuel  power  \
0  Audi   SQ7  2017.0  142800.0           3956.0   Diesel  435.0   
1  Audi    A3  2010.0  236000.0           1395.0  Benzyna  125.0   
2  Audi    A4  2018.0   84500.0           1968.0   Diesel  190.0   
3  Audi    Q5  2018.0   38000.0           1984.0  Benzyna  252.0   
4  Audi    A7  2016.0   74885.0           2995.0  Benzyna  333.0   

        gearbox             drive  city_fuel_consumption  ... doors  seats  \
0  Automatyczna       4x4 (stały)                    8.0  ...   5.0    7.0   
1      Manualna  Na przednie koła                    NaN  ...   5.0    5.0   
2  Automatyczna       4x4 (stały)                    5.0  ...   5.0    5.0   
3  Automatyczna       4x4 (stały)                    8.0  ...   5.0    5.0   
4  Automatyczna       4x4 (stały)                   10.0  ...   5.0    5.0   

    color origin_country accident_free serviced_at_ASO condition  \
0   Szary         Polska           Tak             Tak   Używane   
1  Czarny            NaN           Tak             Tak   Używane   
2   Szary            NaN           NaN             NaN   Używane   
3   Szary            NaN           NaN             NaN   Używane   
4   Biały            NaN           NaN             NaN   Używane   

                price     value  currency  
0  269 000        PLN  269000.0       PLN  
1   29 900        PLN   29900.0       PLN  
2   60 000        PLN   60000.0       PLN  
3   75 000        PLN   75000.0       PLN  
4   75 000        PLN   75000.0       PLN  

[5 rows x 21 columns]

In [1038]:
# Check the rows with EUR currency
eur_price = data[data['currency'] == 'EUR']

print(f'There are {len(eur_price)} rows with EUR currency.')
eur_price.head()

There are 224 rows with EUR currency.


brand         model    year   mileage  engine_capacity         fuel  \
1416  Audi  RS e-tron GT  2023.0      10.0              NaN  Elektryczny   
1431  Audi  RS e-tron GT  2023.0      15.0              NaN  Elektryczny   
3770  Audi            A8  2018.0  114699.0           2967.0       Diesel   
6765  Audi            A4  2020.0   22000.0           1968.0       Diesel   
6786  Audi           RS3  2022.0    6500.0           2480.0      Benzyna   

      power       gearbox        drive  city_fuel_consumption  ... doors  \
1416  598.0  Automatyczna  4x4 (stały)                    NaN  ...   4.0   
1431  598.0  Automatyczna  4x4 (stały)                    NaN  ...   4.0   
3770  286.0  Automatyczna  4x4 (stały)                    8.0  ...   4.0   
6765  190.0  Automatyczna  4x4 (stały)                    5.0  ...   5.0   
6786  400.0  Automatyczna          NaN                   11.1  ...   4.0   

      seats       color origin_country accident_free serviced_at_ASO  \
1416    NaN      Czarny            NaN           NaN             NaN   
1431    NaN     Zielony            NaN           NaN             NaN   
3770    5.0      Czarny         Niemcy           NaN             NaN   
6765    5.0  Inny kolor            NaN           NaN             Tak   
6786    5.0       Szary            NaN           NaN             NaN   

     condition              price    value  currency  
1416   Używane  69 000        EUR  69000.0       EUR  
1431   Używane  70 000        EUR  70000.0       EUR  
3770   Używane  47 000        EUR  47000.0       EUR  
6765   Używane  10 500        EUR  10500.0       EUR  
6786   Używane  55 000        EUR  55000.0       EUR  

[5 rows x 21 columns]

In [1039]:
# Convert EUR to PLN
data.loc[data['currency'] == 'EUR', 'value'] *= 4.45

In [1040]:
# Drop no longer needed columns
data = data.drop(['price', 'currency'], axis='columns')

data = data.rename(columns={"value": "price"})

data.head()

brand model    year   mileage  engine_capacity     fuel  power  \
0  Audi   SQ7  2017.0  142800.0           3956.0   Diesel  435.0   
1  Audi    A3  2010.0  236000.0           1395.0  Benzyna  125.0   
2  Audi    A4  2018.0   84500.0           1968.0   Diesel  190.0   
3  Audi    Q5  2018.0   38000.0           1984.0  Benzyna  252.0   
4  Audi    A7  2016.0   74885.0           2995.0  Benzyna  333.0   

        gearbox             drive  city_fuel_consumption     body  doors  \
0  Automatyczna       4x4 (stały)                    8.0      SUV    5.0   
1      Manualna  Na przednie koła                    NaN  Kompakt    5.0   
2  Automatyczna       4x4 (stały)                    5.0    Kombi    5.0   
3  Automatyczna       4x4 (stały)                    8.0      SUV    5.0   
4  Automatyczna       4x4 (stały)                   10.0  Kompakt    5.0   

   seats   color origin_country accident_free serviced_at_ASO condition  \
0    7.0   Szary         Polska           Tak             Tak   Używane   
1    5.0  Czarny            NaN           Tak             Tak   Używane   
2    5.0   Szary            NaN           NaN             NaN   Używane   
3    5.0   Szary            NaN           NaN             NaN   Używane   
4    5.0   Biały            NaN           NaN             NaN   Używane   

      price  
0  269000.0  
1   29900.0  
2   60000.0  
3   75000.0  
4   75000.0

## **Categorical columns**

To ensure robust analysis, a minimum of 50 observations is set as a benchmark for each category. Any rows that fall short of this criterion will be filtered out, guaranteeing the reliability of the dataset.

Categorical columns in the dataset:
* brand
* model
* fuel_type
* gearbox
* drive
* body_type
* doors
* seats
* color
* origin_country
* accident_free
* serviced_at_ASO
* condition

In [1041]:
data.head()

brand model    year   mileage  engine_capacity     fuel  power  \
0  Audi   SQ7  2017.0  142800.0           3956.0   Diesel  435.0   
1  Audi    A3  2010.0  236000.0           1395.0  Benzyna  125.0   
2  Audi    A4  2018.0   84500.0           1968.0   Diesel  190.0   
3  Audi    Q5  2018.0   38000.0           1984.0  Benzyna  252.0   
4  Audi    A7  2016.0   74885.0           2995.0  Benzyna  333.0   

        gearbox             drive  city_fuel_consumption     body  doors  \
0  Automatyczna       4x4 (stały)                    8.0      SUV    5.0   
1      Manualna  Na przednie koła                    NaN  Kompakt    5.0   
2  Automatyczna       4x4 (stały)                    5.0    Kombi    5.0   
3  Automatyczna       4x4 (stały)                    8.0      SUV    5.0   
4  Automatyczna       4x4 (stały)                   10.0  Kompakt    5.0   

   seats   color origin_country accident_free serviced_at_ASO condition  \
0    7.0   Szary         Polska           Tak             Tak   Używane   
1    5.0  Czarny            NaN           Tak             Tak   Używane   
2    5.0   Szary            NaN           NaN             NaN   Używane   
3    5.0   Szary            NaN           NaN             NaN   Używane   
4    5.0   Biały            NaN           NaN             NaN   Używane   

      price  
0  269000.0  
1   29900.0  
2   60000.0  
3   75000.0  
4   75000.0

### **brand**

In [1042]:
data['brand'].value_counts()

BMW              17721
Audi             17581
Volkswagen       17505
Ford             15372
Opel             14518
Mercedes-Benz    13599
Renault          10258
Toyota            9833
Skoda             9511
Peugeot           8155
Seat                 2
Lancia               1
Name: brand, dtype: int64

In [1043]:
# Brands Seat and Lancia should not be in the dataset, so I drop them
data = data[~data['brand'].isin(['Seat', 'Lancia'])]

### **model**

In [1044]:
data.groupby(['brand', 'model']).size()

brand          model                                                                                 
Audi           100                                                                                         25
               200                                                                                          2
               80                                                                                          65
               90                                                                                           1
               A1                                                                                         206
               A2                                                                                          38
               A3                                                                                        2344
               A4                                                                                        4136
               A4 

In [1045]:
value_counts = data['model'].value_counts()
filtered_models = value_counts[value_counts < 50].index
print(f'There are {len(filtered_models)} models with less than 50 offers.')

data = data[~data['model'].isin(filtered_models)]

There are 184 models with less than 50 offers.


### **fuel**

In [1046]:
# Translating fuel types from polish to english
translations = {
    'Diesel': 'Diesel',
    'Benzyna': 'Gasoline',
    'Hybryda': 'Hybrid',
    'Benzyna+LPG': 'Gasoline+LPG',
    'Elektryczny': 'Electric',
    'Benzyna+CNG': 'Gasoline+CNG',
    'Wodór': 'Hydrogen',
    'Etanol': 'Ethanol'
}

data['fuel'] = data['fuel'].replace(translations)

In [1047]:
data['fuel'].value_counts()

Diesel          61639
Gasoline        59413
Hybrid           4217
Gasoline+LPG     3954
Electric         2118
Gasoline+CNG       47
Ethanol             1
Name: fuel, dtype: int64

In [1048]:
# What is interesting, there is one car that is fueled by Ethanol
data[data['fuel'] == 'Ethanol']

brand model    year  mileage  engine_capacity     fuel  power  \
117494  Volkswagen  ID.5  2023.0      7.0           1395.0  Ethanol  204.0   

             gearbox          drive  city_fuel_consumption body  doors  seats  \
117494  Automatyczna  Na tylne koła                    NaN  SUV    5.0    5.0   

        color origin_country accident_free serviced_at_ASO condition     price  
117494  Biały         Polska           Tak             NaN      Nowe  258790.0

In [1049]:
# Filtering out cars with fuel type that do not meet the benchmark
data = data[~data['fuel'].isin(['Gasoline+CNG', 'Ethanol'])]

### **gearbox**

In [1050]:
# Translating gearbox types from polish to english
translations = {
    'Manualna': 'Manual',
    'Automatyczna': 'Automatic',
}

data['gearbox'] = data['gearbox'].replace(translations)

In [1051]:
data['gearbox'].value_counts()

Manual       71311
Automatic    59976
Name: gearbox, dtype: int64

### **drive**

In [1052]:
# Translating drive types from polish to english
translations = {
    'Na przednie koła': 'Front-wheel drive',
    '4x4 (stały)': 'All-wheel drive (permanent)',
    'Na tylne koła': 'Rear-wheel drive',
    '4x4 (dołączany automatycznie)': 'Four-wheel drive (automatic engagement)',
    '4x4 (dołączany ręcznie)': 'Four-wheel drive (manual engagement)'
}

data['drive'] = data['drive'].replace(translations)

In [1053]:
data['drive'].value_counts()

Front-wheel drive                          74323
All-wheel drive (permanent)                13540
Rear-wheel drive                           13177
Four-wheel drive (automatic engagement)    12300
Four-wheel drive (manual engagement)         853
Name: drive, dtype: int64

### **body**

In [1054]:
# Translating body types from polish to english
translations = {
    'Kombi': 'Estate',
    'SUV': 'SUV',
    'Sedan': 'Sedan',
    'Kompakt': 'Compact',
    'Minivan': 'Minivan',
    'Auta miejskie': 'City_car',
    'Coupe': 'Coupe',
    'Auta małe': 'Small_car',
    'Kabriolet': 'Convertible'
}

data['body'] = data['body'].replace(translations)

In [1055]:
data['body'].value_counts()

Estate         29889
SUV            26222
Sedan          22624
Compact        20877
Minivan        12568
City_car        8201
Coupe           5365
Small_car       3381
Convertible     2214
Name: body, dtype: int64

### **doors**

In [1056]:
data['doors'].value_counts()

5.0    102121
4.0     17541
3.0      6068
2.0      4837
6.0        55
Name: doors, dtype: int64

### **seats**

In [1057]:
data['seats'].value_counts()

5.0    107569
4.0      7265
7.0      4922
2.0      1153
9.0       816
6.0       508
8.0       466
3.0       430
1.0        34
Name: seats, dtype: int64

In [1058]:
# Filtering out cars number of seats that do not meet the benchmark
data = data[data['seats'] != 1]

### **color**

In [1059]:
# Translating colors from polish to english
translations = {
    'Czarny': 'Black',
    'Szary': 'Gray',
    'Biały': 'White',
    'Srebrny': 'Silver',
    'Niebieski': 'Blue',
    'Inny kolor': 'Other_color',
    'Granatowy': 'Navy_blue',
    'Czerwony': 'Red',
    'Brązowy': 'Brown',
    'Zielony': 'Green',
    'Bordowy': 'Burgundy',
    'Złoty': 'Gold',
    'Beżowy': 'Beige',
    'Błękitny': 'Light_blue',
    'Pomarańczowy': 'Orange',
    'Żółty': 'Yellow',
    'Fioletowy': 'Purple'
}

data['color'] = data['color'].replace(translations)

In [1060]:
data['color'].value_counts()

Black          34105
Gray           21378
White          21052
Silver         18501
Blue            9403
Other_color     5695
Navy_blue       4895
Red             4768
Brown           3031
Green           1942
Burgundy        1522
Gold            1376
Beige           1173
Light_blue      1109
Orange           530
Yellow           469
Purple           358
Name: color, dtype: int64

### **origin_country**

In [1061]:
# Translating origin countries from polish to english
translations = {
    'Polska': 'Poland',
    'Niemcy': 'Germany',
    'Francja': 'France',
    'Stany_Zjednoczone': 'United_States',
    'Belgia': 'Belgium',
    'Szwajcaria': 'Switzerland',
    'Holandia': 'Netherlands',
    'Włochy': 'Italy',
    'Szwecja': 'Sweden',
    'Austria': 'Austria',
    'Dania': 'Denmark',
    'Kanada': 'Canada',
    'Japonia': 'Japan',
    'Czechy': 'Czech_Republic',
    'Wielka_Brytania': 'United_Kingdom',
    'Luksemburg': 'Luxembourg',
    'Norwegia': 'Norway',
    'Hiszpania': 'Spain',
    'Inny': 'Other',
    'Finlandia': 'Finland',
    'Słowacja': 'Slovakia',
    'Słowenia': 'Slovenia',
    'Grecja': 'Greece',
    'Litwa': 'Lithuania',
    'Estonia': 'Estonia',
    'Węgry': 'Hungary',
    'Rumunia': 'Romania',
    'Irlandia': 'Ireland',
    'Islandia': 'Iceland',
    'Liechtenstein': 'Liechtenstein',
    'Łotwa': 'Latvia',
    'Rosja': 'Russia',
    'Monako': 'Monaco',
    'Bułgaria': 'Bulgaria',
    'Ukraina': 'Ukraine',
    'Chorwacja': 'Croatia',
    'Turcja': 'Turkey',
    'Korea': 'Korea'
}

data['origin_country'] = data['origin_country'].replace(translations)

In [1062]:
data['origin_country'].value_counts()

Poland               42241
Germany              27324
France                4411
Stany Zjednoczone     2864
Belgium               2790
Switzerland           1515
Netherlands           1202
Italy                  915
Sweden                 758
Austria                670
Denmark                459
Canada                 426
Japan                  301
Czech_Republic         178
Wielka Brytania        143
Luxembourg             127
Norway                 113
Spain                   89
Other                   73
Finland                 67
Slovakia                32
Slovenia                 6
Greece                   6
Lithuania                5
Estonia                  5
Hungary                  5
Romania                  3
Ireland                  3
Iceland                  3
Liechtenstein            3
Latvia                   2
Russia                   2
Monaco                   1
Bulgaria                 1
Ukraine                  1
Croatia                  1
Turkey                   1
K

In [1063]:
value_counts = data['origin_country'].value_counts()
filtered_countries = value_counts[value_counts < 50].index
print(f'There are {len(filtered_countries)} countries of origin with less than 50 offers.')

data = data[~data['origin_country'].isin(filtered_countries)]

There are 18 countries of origin with less than 50 offers.


### **accident_free**

In [1064]:
# Translating whether the car is accident free from polish to english
data.loc[data['accident_free'] == 'Tak', 'accident_free'] = True

In [1065]:
data['accident_free'].value_counts()

True    74029
Name: accident_free, dtype: int64

In [1066]:
# I will assume that if someone has not marked that the car is accident free, then it is not
data['accident_free'] = data['accident_free'].fillna(False)

In [1067]:
data['accident_free'].value_counts()

True     74029
False    57197
Name: accident_free, dtype: int64

### **serviced_at_ASO**

In [1068]:
# Translating whether the car was serviced at ASO from polish to english
data.loc[data['serviced_at_ASO'] == 'Tak', 'serviced_at_ASO'] = True

In [1069]:
data['serviced_at_ASO'].value_counts()

True    59619
Name: serviced_at_ASO, dtype: int64

In [1070]:
# I will assume that if someone has not marked that the car was serviced at ASO, then it was not
data['serviced_at_ASO'] = data['serviced_at_ASO'].fillna(False)

In [1071]:
data['serviced_at_ASO'].value_counts()

False    71607
True     59619
Name: serviced_at_ASO, dtype: int64

### **condition**

In [1072]:
# Translating car condition from polish to english
translations = {
    'Używane': 'Used',
    'Nowe': 'New'
}

data['condition'] = data['condition'].replace(translations)

In [1073]:
data['condition'].value_counts()

Used    119430
New      11796
Name: condition, dtype: int64

## **Saving cleaned dataset**

In [1076]:
data.to_csv('data_scraping/data/all_data_cleaned.csv', index=False)